### PR TITLE
Port redraw-frame, redraw-display, and associated C functions

### LINKED ISSUE #929: Port redraw-frame
```c
DEFUN ("redraw-frame", Fredraw_frame, Sredraw_frame, 0, 1, 0,
       doc: /* Clear frame FRAME and output again what is supposed to appear on it.
If FRAME is omitted or nil, the selected frame is used.  */)
  (Lisp_Object frame)
{
  redraw_frame (decode_live_frame (frame));
  return Qnil;
}
```

### PR DESCRIPTION
**Pull request by bb010g**
I'm not too sure if I got the `unsafe fn` attributes right.

Closes #929.

---
This change is [Reviewable](https://reviewable.io/reviews/wilfred/remacs/962)

### CODE REVIEW

File: rust_src/src/dispnew.rs [hunk 1]
```diff
@@ -5,10 +5,15 @@ use std::{cmp, ptr};
 use remacs_lib::current_timespec;
 use remacs_macros::lisp_fn;
 use remacs_sys::Qnil;
```
Comment (shaleh) [inline rust_src/src/dispnew.rs]: Typically we group functions together and group types.

File: rust_src/src/dispnew.rs [hunk 2]
```diff
@@ -38,4 +43,64 @@ pub fn sleep_for(seconds: EmacsDouble, milliseconds: Option<EmacsInt>) -> () {
...
+/// Clear frame FRAME and output again what is supposed to appear on it.
+/// If FRAME is omitted or nil, the selected frame is used.
+#[lisp_fn(c_name = "redraw_frame", name = "redraw-frame", min = "0")]
+pub fn redraw_frame_lisp(frame: LispObject) {
+    unsafe { redraw_frame(frame_live_or_selected(frame).as_mut()) };
+}
+
+/// Clear and redisplay all visible frames.
+#[lisp_fn]
+pub fn redraw_display() {
```
Comment (shaleh) [inline rust_src/src/dispnew.rs]: Worth switching to `for_each` here?

**Comment by bb010g** [comment]
I forgot to save and commit my clearing out of stuff from `dispextern.h`, but trying to figure out what makes cc happy is taking a while due to build times.

File: rust_src/src/terminal.rs [hunk 3]
```diff
@@ -0,0 +1,83 @@
+use libc::c_int;
+use remacs_sys::{frame, glyph};
+
+#[no_mangle]
+pub unsafe extern "C" fn update_begin(f: *mut frame) {
```
Comment (harryfei) [inline rust_src/src/terminal.rs]: We can use `LispFrameRef` defined in `frames.rs`. The code can be like this.
```rust
#[no_mangle]
pub extern "C" fn update_begin(f: LispFrameRef) {
   if let Some(hook) = (*(f.terminal)).update_begin_hook {
        hook(f.as_mut())
    }
}
```

if we defined a new `TerminalRef` type just as the `LispFrameRef`, the above code could be more rusty. All the magic are from `ExternalPtr` in lisp.rs. You can check out it.

Comment (harryfei) [reply rust_src/src/terminal.rs]: `*mut window` is just same.

Comment (harryfei) [reply rust_src/src/terminal.rs]: Same with `*mut window`.

Comment (brotzeit) [reply rust_src/src/terminal.rs]: You can copy it from #871

Comment (shaleh) [reply rust_src/src/terminal.rs]: @harryfei this is a C exposed function not a `lisp_fn`.

Comment (bb010g) [reply rust_src/src/terminal.rs]: Is it safe to use that here? It's `#[repr(C)]`, not `#[repr(transparent)]`.

Comment (shaleh) [reply rust_src/src/terminal.rs]: @bb010g look at https://github.com/Wilfred/remacs/blob/a613f9ff65ad7ef039c26c403302e2f7e4b5410f/rust_src/src/windows.rs#L652 then look at `LispWindowRef`

**Pull request merged**